### DISCUSSION TITLE
Help with flashing serial firmware

### DISCUSSION THREAD

**Discussion by PaulAntonDeen** [gcormier/megadesk#101] [Q&A] [answered]
First time working with avrdude here. I need a little help, I understand that if there is a bootloader on this thing than I can flash with an FTDI cable...is there?
If not, I need an AVR programmer: i just ordered an USBasp to be sure.
Is there a little tutorial using windows or mac on how to flash the .hex file for serial communication? I see way too many options and documentation online. I don't fully understand what fuses are yet and if they need to be done in seperate commands or can be included in the flash command. So some orientation is much appreciated!

**Comment by gcormier** [comment]
There is no bootloader so you need an AVR programmer.

Some info here
https://github.com/gcormier/megadesk/discussions/49

Also here
https://github.com/gcormier/megadesk/blob/master/DIY.md

Windows sample file to flash
https://github.com/gcormier/megadesk/blob/master/flash.bat

I do need to upload my linux script, but this will get you started, this is for the Pololu AVR programmer. If you comment/enable the respective lines it should work
```
#!/bin/bash
# Pololu
PORT=`pavr2cmd --prog-port`
PROG=stk500v2
pavr2cmd --freq 114
# End Pololu

# avrisp
#PROG=usbtiny
#PORT=usb
# end avrisp


MCU=attiny841

while  :
do
pavr2cmd --freq 114
avrdude -c $PROG -p $MCU -U lfuse:w:0xe2:m -P $PORT
avrdude -c $PROG -p $MCU -U hfuse:w:0xd6:m -P $PORT
avrdude -c $PROG -p $MCU -U efuse:w:0xfe:m -P $PORT

avrdude -B 0.5 -c $PROG -p $MCU -V -Uflash:w:megadesk-v2022.03.hex:i -P $PORT  

read -n 1 -s -r -p "Press any key to continue"

done
```

**Reply by PaulAntonDeen** [reply]
Thanks I'll give it a shot and close this one if I can make it work :)

**Reply by gcormier** [reply]
Sounds good let me know if you have any issues.

**Reply by PaulAntonDeen** [reply]
Got Zadig to install the device driver for the programmer: https://quadmeup.com/working-solution-for-usbasp-driver-in-windows-10-64bit/
Connected all MOSI - MOSI, MISO-MISO, SCK-SCK, RST-RST and 5V and gnd. I get the beeping board so that sounds like it went well.
Then using your bat file (adjusted the path to conf) I get one shot to set a fuse and then it doesn't respond anymore after that. Any suggestions?

![image](https://user-images.githubusercontent.com/25848961/163252256-8b30c761-642f-49ed-9fc5-b5050677a8d0.png)

**Comment by PaulAntonDeen** [comment]
I think I need to do something with the clock speed according to (#49)?
-B 3 according to https://electronics.stackexchange.com/questions/49606/avrdude-b-3-parameter...but it doesn't work :(

**Reply by gcormier** [reply]
You can try using -B to slow it down since it's only running on the internal 8MHz oscillator. I think I used to use `-B 0.5`? But otherwise I have never had much luck with Zadig and/or the cheaper programmers that need it.

**Comment by PaulAntonDeen** [comment, marked as answer]
Turns out I could set 1 fuse and then nothing else...but that is remembered at reboot. So I just unplugged the USBasp and plugged it back in and proceeded to the second fuse, the third and finally the flash. It works now :) I ended up doing the whole thing on a rPi4 under rasbian OS and used your second script...but manually typed.

**Reply by gcormier** [reply]
Sounds about right for the inconsistent and frustrating experiences to be had with usbasp :) Glad you got it going!

**Comment by farsonic** [comment]
I'm trying to update the flash on my megadesk to provide the serial support needed for megadesk. I've tried using a buspirate as well as a usbASP to do the programing and I've soldered header pins on the the megaboard (v4) also to make this simpler. I've validated the usbASP is functional by programming using the same header into an old Arduino Mega Pro. 

I've tried all combinations/variations to even set the fuses ... this is the error I'm receiving. I'm going to try another computer next. 

>>>: avrdude -u -c usbasp -p t841 -P usb -B 1500.0 -F -v -U lfuse:w:0xd6:m 

avrdude.exe: Version 6.3, compiled on Feb 17 2016 at 09:25:53
             Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
             Copyright (c) 2007-2014 Joerg Wunsch

             System wide configuration file is "C:\Program Files (x86)\AVRDUDESS\avrdude.conf"

             Using Port                    : usb
             Using Programmer              : usbasp
             Setting bit clk period        : 1500.0
             AVR Part                      : ATtiny841
             Chip Erase delay              : 15000 us
             PAGEL                         : P00
             BS2                           : P00
             RESET disposition             : possible i/o
             RETRY pulse                   : SCK
             serial program mode           : yes
             parallel program mode         : yes
             Timeout                       : 200
             StabDelay                     : 100
             CmdexeDelay                   : 25
             SyncLoops                     : 32
             ByteDelay                     : 0
             PollIndex                     : 3
             PollValue                     : 0x53
             Memory Detail                 :

                                      Block Poll               Page                       Polled
               Memory Type Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
               ----------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
               eeprom        65     6     4    0 no        512    4      0  4000  4500 0xff 0xff
               flash         65    12    32    0 yes      8192   16    512  4500  4500 0xff 0xff
               signature      0     0     0    0 no          3    0      0     0     0 0x00 0x00
               lock           0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
               lfuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
               hfuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
               efuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
               calibration    0     0     0    0 no          1    0      0     0     0 0x00 0x00

             Programmer Type : usbasp
             Description     : USBasp, http://www.fischl.de/usbasp/

avrdude.exe: set SCK frequency to 500 Hz
avrdude.exe: warning: cannot set sck period. please check for usbasp firmware update.
avrdude.exe: error: program enable: target doesn't answer. 1 
avrdude.exe: initialization failed, rc=-1
avrdude.exe: AVR device initialized and ready to accept instructions
avrdude.exe: Device signature = 0x743230
avrdude.exe: Expected signature for ATtiny841 is 1E 93 15

**Reply by farsonic** [reply]
can someone confirm the orientation of the headers for the connection for the AVR programmer...the picture on the DIY page doesn't state the orientation of the connector or which side it is interfacing from.

**Reply by farsonic** [reply]
ok, looks like a set the lfuse incorrectly and can no longer program the 841 on the megadesk. I've managed to hook up an arduino in an attempt to program it using the high voltage approach. However, can I program this while it is installed on the board? I'm having to run this arduino sketch https://www.rickety.us/2010/03/arduino-avr-high-voltage-serial-programmer/comment-page-2 and all I've done is change the lfuse settings etc. It runs and seems to get some communication with the chip but not making any changes. Here is the output...

10
Entering programming Mode

lfuse reads as EC
hfuse reads as FE
efuse reads as EE

Writing hfuse
Writing lfuse

lfuse reads as EC
hfuse reads as FE
efuse reads as EE

Anyone got any suggestions?

**Reply by gcormier** [reply]
Removing the IC is probably safest, but will require soldering skills. I'm not sure on the high voltage programming method and if it is safe to do so while it's in-circuit.

**Comment by farsonic** [comment]
Looks like I’ll be ordering another unit…

Might see about ordering another tiny 841 for this one for potential recovery also. 

Re

Francois 

19 Oct 2023 at 06:10 by ***@***.***:
